Determine component compatibility with MeiliSearch (#417)
meilisearch/instant-meilisearch: README.md

NB: If you don't have any MeiliSearch instance running and containing your data, you should take a look at this [getting started page](https://docs.meilisearch.com/learn/tutorials/getting_started.html).

## Table of Contents <!-- omit in toc -->

- [🔧 Installation](#-installation)
- [🎬 Usage](#-usage)
- [⚡️ Example with InstantSearch](#-example-with-instantSearch)

> - [🤖 Compatibilities](#-compatibility-with-meilisearch-and-instantsearch)

Reviewer: ```suggestion
- [🤖 Compatibility with MeiliSearch and InstantSearch](#-compatibility-with-meilisearch-and-instantsearch)
```

- If you use Vue, check out [meilisearch-vue](https://github.com/meilisearch/meilisearch-vue/)
- If you use Angular, check out [meilisearch-angular](https://github.com/meilisearch/meilisearch-angular/)

## API resources

List of all the components that are available in [instantSearch](https://github.com/algolia/instantsearch.js) and their compatibilty with [MeiliSearch](https://github.com/meilisearch/meilisearch-react/).

### ✅ InstantSearch

[instantSearch references](https://www.algolia.com/doc/api-reference/widgets/instantsearch/js/)

`instantSearch` is the main component. It manages the widget and lets you add new ones.

> - ✅ IndexName: Name of your index. _required_

Reviewer: Can you precise it corresponds to the Index UID for MeiliSearch?

- ✅ SearchClient: Search client, in our case instantMeiliSearch. See [customization](#customization) for details on options. _required_

> - ❌ numberLocale: Did not work with both algoliasearch and instantMeiliSearch.

Reviewer: ```suggestion
- ❌ numberLocale: Does not work with both Algoliasearch and InstantMeiliSearch.
```

Reviewer: "Did" is confusing, the users don't have the context of your investigation when reading the README.

- ✅ searchFunction: Surcharge the search function provided by the search client.
- ✅ initialUiState: Determine the search state on app start.
- ✅ onStateChange: Change search state on change (see option above).
- ✅ stalledSearchDelay: Time in ms before search is considered stalled. [Used for loader](https://www.algolia.com/doc/guides/building-search-ui/going-further/improve-performance/js/#using-the-searchbox).
- ✅ routing: browser URL synchronization, search parameters appear in current URL ([guide](https://www.algolia.com/doc/guides/building-search-ui/going-further/routing-urls/js/)).
- ✅ insightsClient: Hook analytics to search actions ([see insight section](#insight)).

```js
const search = instantsearch({
  indexName: 'instant_search',
  searchClient: instantMeiliSearch(
    'https://demos.meilisearch.com',
    'dc3fedaf922de8937fdea01f0a7d59557f1fd31832cb8440ce94231cfdde7f25',
    {...instantMeiliSearchOptions}
  ),

>   {...instantSearchOptions}

Reviewer: Why `{}` around? It looks like `instantsearch` is defined like this:
```js
const search = instantsearch({
  indexName: '...',
  searchClient: ...,
  option1: ...,
  option2: ...,
  ...
});
```

and not:

```js
const search = instantsearch({
  indexName: '...',
  searchClient: ...,
  {
    option1: ...,
    option2: ...,
    ...
  }
});
```

})
```

### ❌ Index

[Index references](https://www.algolia.com/doc/api-reference/widgets/index-widget/js/)

`Index` is the component that lets you apply widgets to a dedicated index. It’s useful if you want to build an interface that targets multiple indices.

Not compatible as MeiliSearch does not support federated search on multiple indexes.

If you'd like to see federated search implemented please vote for it in the [roadmap](https://roadmap.meilisearch.com/c/74-multi-index-search?utm_medium=social&utm_source=portal_share).

### ✅ SearchBox

[SearchBox references](https://www.algolia.com/doc/api-reference/widgets/search-box/js/)

The `searchBox` widget is used to let the user perform a text-based query.

- ✅ container: The CSS Selector or HTMLElement to insert the widget into. _required_
- ✅ placeholder: Placeholder of the search box.

> - ✅ autofocus: Wether the search box is focused on arrival.

Reviewer: ```suggestion
- ✅ autofocus: Whether the search box is focused on arrival.
```

> - ✅ searchAsYouType: Whether result appear as you type or after pressing enter.

Reviewer: ```suggestion
- ✅ searchAsYouType: Whether result appears as you type or after pressing enter.
```

> - ❌ showReset: Did not work with both algoliaSearch and instantMeiliSearche
> - ❌ showSubmit: Did not work with both algoliaSearch and instantMeiliSearch

Reviewer: ```suggestion
- ❌ showReset: Does not work with both algoliaSearch and instantMeiliSearche
- ❌ showSubmit: Does not work with both algoliaSearch and instantMeiliSearch
```

- ✅ showLoadingIndicator: Whether to show the spinning loader.
- ✅ queryHook: A function that is called just before the search is triggered.
- ✅ templates: The templates to use for the widget.
- ✅ cssClasses: The CSS classes to override.

```js
instantsearch.widgets.searchBox({
  container: '#searchbox',
  autofocus: true,
  ...searchBoxOptions
}),
```

### ✅ Configure

- [Configure references](https://www.algolia.com/doc/api-reference/widgets/configure/js/): See #389

The `configure` widget lets you provide raw search parameters to the Algolia API without rendering anything.

> Because these are the search parameters of AlgoliaSearch and not the InstantSearch parameters, some of them are ignored by InstantSearch.

Reviewer: ```suggestion
Because these are the search parameters of AlgoliaSearch and not the InstantSearch parameters, some of them are ignored by InstantSearch.<br>
```

> Since we do not act as AlgoliaSearch on search parameters detailed compatibility can be found in [this issue](https://github.com/meilisearch/instant-meilisearch/issues/389).

Reviewer: ```suggestion
Since we do not act as AlgoliaSearch on search parameters, detailed compatibility can be found in [this issue](https://github.com/meilisearch/instant-meilisearch/issues/389).<br>
```

### Panel

[Panel references](https://www.algolia.com/doc/api-reference/widgets/panel/js/)

The `panel` widget wraps other widgets in a consistent panel design.

- ✅ hidden: Function to determine if pannel should be hidden on earch render.
- ✅ collapsed: Function to determine if pannel should be collapsed on earch render.
- ✅ templates: The templates to use for the widget.
- ✅ cssClasses: The CSS classes to override.

### ❌ Autocomplete

[Autocomplete references](https://www.algolia.com/doc/api-reference/widgets/autocomplete/js/)

Deprecated component in InstantSearch in favor of [autocomplete package](https://github.com/algolia/autocomplete).

InstantMeiliSearch is not compatible with the autocomplete package.

### ✅ Voice Search

[Voice Search references](https://www.algolia.com/doc/api-reference/widgets/voice-search/js/)

The `voiceSearch` widget lets the user perform a voice-based query.

### ✅ Insight

[Insight references](https://www.algolia.com/doc/api-reference/widgets/insights/js/)

Search Insights lets you report click, conversion and view metrics.

More details about the subject are [given in this issue](https://github.com/meilisearch/instant-meilisearch/issues/410).

> Require InstantSearch v4.8.3 or later

Reviewer: ```suggestion
Requires InstantSearch v4.8.3 or later.
```

> - ✅ insightsClient: Insight Client uses [search-insight.js](https://github.com/algolia/search-insights.js).

Reviewer: ```suggestion
- ✅ insightsClient: Insight client uses [search-insight.js](https://github.com/algolia/search-insights.js).
```

Used to display a list of results.

- ✅ container: The CSS Selector or HTMLElement to insert the hits into. _required_
- ✅ escapeHTML: Escapes HTML tags in string values in returned hits.
- ✅ transformItems: Function that maps over every hit the provided logic.
- ✅ templates: The templates to use for the widget.
- ✅ cssClasses: The CSS classes to override.

```js
instantsearch.widgets.hits({
  container: '#hits',
  templates: {
    item: `
        <div>
          <div class="hit-name">
            {{#helpers.highlight}}{ "attribute": "title" }{{/helpers.highlight}}
          </div>
        </div>
      `,
  },
})
```

### ✅ InfiniteHits

[infiniteHits references](https://www.algolia.com/doc/api-reference/widgets/infinite-hits/js/)

The `infiniteHits` widget is used to display a list of results with a “Show more” button.

- ✅ container: The CSS Selector or HTMLElement to insert the hits into. _required_
- ✅ escapeHTML: Escapes HTML tags in string values in returned hits.
- ✅ transformItems: Function that maps over every hit the provided logic.
- ✅ templates: The templates to use for the widget.
- ✅ cssClasses: The CSS classes to override.

> - ❌ showPrevious: Did not work with both Algoliasearch and InstantMeiliSearch.

Reviewer: ```suggestion
- ❌ showPrevious: Does not work with both Algoliasearch and InstantMeiliSearch.
```

- ❌ cache: Not added in InstantMeiliSearch.

```js
instantsearch.widgets.infiniteHits({
  container: '#infinite-hits',
  templates: {
    item: `
      <h2>
        {{#helpers.highlight}}{ "attribute": "name" }{{/helpers.highlight}}
      </h2>
    `,
  },
})
```

### ✅ Highlight

[highlight references](https://www.algolia.com/doc/api-reference/widgets/highlight/js/)

The `highlight` function returns an attribute from a hit into its highlighted form, when relevant.

- ✅ attribute: The attribute of the record to highlight. _required_
- ✅ hit: Hit object. _required_
- ✅ highlightedTagName: HTML element to wrap the highlighted parts of the string.

See [Hits](#hits) for an example.

> ### ✅ Snippet

Reviewer: Can you precise it corresponds to the "Cropping"  of MeiliSearch?

Reviewer: And the crop in MeiliSearch is in char, not in word like Algolia does.
`["overview:20"]` is not the same as with Algolia.

Author: It works the same way with algolia I think: https://codesandbox.io/s/jolly-hypatia-xtkv1?file=/src/App.js
Could you confirm with this codesandbox ?

Reviewer: - change `description:20` by `description:2` -> you will get 2 words, and only 1 with MeiliSearch
- https://www.algolia.com/doc/guides/building-search-ui/ui-and-ux-patterns/highlighting-snippeting/js/#nbwords

Author: Ah I thought you talked about usage and not how features works with algolia's compared to meilisearch.

The difference happens in algolia's core. See [this issue]((https://github.com/algolia/react-instantsearch/issues/3046#issuecomment-827719846)) that I think confirms that this happens in their core. and is thus may not be related to compatibility with`instantSearch.js` components. 

Should we still document this here or make a link to the issue that details the differences between `meilisearch` and `algoliasearch` (see #389)?

Maybe we could say something like 
```
Note that `snippeting` works differently with the `algoliasearch-client-javascript` client than with `instant-meilisearch` client. Precise comparison is [given in this issue](https://github.com/meilisearch/instant-meilisearch/issues/389).
```

> [Snippet references](https://www.algolia.com/doc/api-reference/widgets/snippet/react/)

Reviewer: ```suggestion
[Snippet references](https://www.algolia.com/doc/api-reference/widgets/snippet/js/)
```

- ✅ hit: Hit object. _required_
- ✅ highlightedTagName: HTML element to wrap the highlighted parts of the string.

Note that the attribute has to be added to `attributesToSnippet` in [configuration](#configuration). Highlight is applied on snippeted fields.

```js
instantsearch.widgets.configure({
  attributesToSnippet: ['description'],
})
```

```js
instantsearch.widgets.hits({
  // ...
  templates: {
    item: `
      <p>{{#helpers.snippet}}{ "attribute": "description" }{{/helpers.snippet}}</p>
    `,
  },
})
```

### ❌ Geo Search

[Geo search references](https://www.algolia.com/doc/api-reference/widgets/geo-search/js/)

No compatibility because MeiliSearch does not support Geo Search.

If you'd like to see it implemented please vote for it in the [roadmap](https://roadmap.meilisearch.com/c/33-geo-search?utm_medium=social&utm_source=portal_share).

### ❌ Answers

[Answers references](https://www.algolia.com/doc/api-reference/widgets/answers/js/).

No compatibility because MeiliSearch does not support this experimental feature.

### ✅ RefinementList

[Refinement list references](https://www.algolia.com/doc/api-reference/widgets/refinement-list/js/)

> The `refinementList` widget is one of the most common widget you can find in a search UI. With this widget, the user can filter the dataset based on facets.

Reviewer: ```suggestion
The `refinementList` widget is one of the most common widgets you can find in a search UI. With this widget, the user can filter the dataset based on facets.
```

- ✅ container: The CSS Selector or HTMLElement to insert the refinements. _required_
- ✅ attribute: The facet to display _required_
- ✅ operator: How to apply facets, "AND" or "OR"
- ✅ limit: How many facet values to retrieve.
- ✅ showMore: Whether to display a button that expands the number of items.
- ✅ showMoreLimit: The maximum number of displayed items. Does not work when showMoreLimit > limit.
- ❌ searchable: Whether to add a search input to let the user search for more facet values. Not supported by MeiliSearch. If you'd like to see it implemented [please vote](https://roadmap.meilisearch.com/c/64-search-for-facet-values?utm_medium=social&utm_source=portal_share).

> - ❌ searchablePlaceholder: The value of the search input’s placeholder, not supported, see `searchable`.
> - ❌ searchableIsAlwaysActive: When false, disables the facet search input, not supported, see `searchable`.
> - ❌ searchableEscapeFacetValues: When true, escapes the facet values, not supported, see `searchable`.
> - ❌ sortBy: Not supported natively but can be implemented manually using `transformItems` options.

Reviewer: ```suggestion
- ❌ searchablePlaceholder: The value of the search input’s placeholder. Not supported, see `searchable`.
- ❌ searchableIsAlwaysActive: When false, disables the facet search input. Not supported, see `searchable`.
- ❌ searchableEscapeFacetValues: When true, escapes the facet values. Not supported, see `searchable`.
- ❌ sortBy: Not supported natively but can be implemented manually using `transformItems` options.
```

No compatibility because MeiliSearch does not support hierarchical facets.

If you'd like get nested facets/hierarchical facets implemented, please vote for it in the [roadmap](https://roadmap.meilisearch.com/c/97-nested-facets?utm_medium=social&utm_source=portal_share).

### ✅ RangeSlider

[Range slider references](https://www.algolia.com/doc/api-reference/widgets/range-slider/js/)

The `rangeSlider` widget provides a user-friendly way to filter the results, based on a single numeric range.

- ✅ container: The CSS Selector or HTMLElement to insert the refinements. _required_
- ✅ attribute: The name of the attribute in the document. _required_.
- ✅ min: The minimum value for the input. _required_
- ✅ max: The maximum value for the input. _required_
- ❌ precision: The number of digits after the decimal point to use. Not compatible as only integers work with `rangeSlider`.
- ✅ step: The number of steps between each handle move.
- ✅ pips: Whether to show slider pips (ruler marks).
- ✅ tooltips: Whether to show tooltips. The default tooltips show the raw value.
- ✅ cssClasses: The CSS classes to override.

> ⚠️ The component is compatible but requires the following workaround:https://www.algolia.com/doc/api-reference/widgets/js/

Reviewer: ```suggestion
⚠️ The component is compatible but requires the following workaround: https://www.algolia.com/doc/api-reference/widgets/js/
```

Reviewer: I don't get which workaround we need to do

Reviewer: Do you mean the next step following? It looks like the workaround is in the link you put after the sentence

#### Only integers

Only attributes with integers values are compatible.

#### Manual Min Max

Min and max of attributes are not returned from MeiliSearch and thus **must be set manually**.

```js
  instantsearch.widgets.rangeSlider({
    // ...
    min: 0,
    max: 100000,
  }),
```

#### Remove attributes from facet list

If the attribute is not in the [`attributesForFacetting`](https://docs.meilisearch.com/reference/features/settings.html#attributes-for-faceting) setting list, an error is thrown.
The error comes from InstantSearch adding the attribute in the facets list and MeiliSearch throwing an error when a non-existent facet is queried on search.

```js
  instantsearch.widgets.rangeSlider({
    attribute: 'recommendationCount',
    // ...
  }),
```

Throws the following error.

```json
{
  "message": "Attribute recommendationCount is not set as facet",
  "errorCode": "bad_request",
  "errorType": "invalid_request_error",
  "errorLink": "https://docs.meilisearch.com/errors#bad_request"
}
```

> To avoid this error, the facet should be removed from the query parameter before it is send to MeiliSearch.

Reviewer: ```suggestion
To avoid this error, the facet should be removed from the query parameter before it is sent to MeiliSearch.
```

The `breadcrumb` widget is a secondary navigation scheme that lets the user see where the current page is in relation to the facet’s hierarchy.

No compatibility because MeiliSearch does not support hierarchical facets.

If you'd like get nested facets implemented, please vote for it in the [roadmap](https://roadmap.meilisearch.com/c/97-nested-facets?utm_medium=social&utm_source=portal_share).

### ✅ Stats

[Stats references](https://www.algolia.com/doc/api-reference/widgets/stats/js/)

The `stats` widget displays the total number of matching hits and the time it took to get them (time spent in the Algolia server).

- ✅ container: The CSS Selector or HTMLElement to insert the widget into. _required_
- ✅ cssClasses: The CSS classes to override.
- ✅ transformItems: function receiving the items, called before displaying them.

```js
instantsearch.widgets.stats({
  container: '#stats',
})
```

### ❌ Analytics

[Analytics](https://www.algolia.com/doc/api-reference/widgets/analytics/js/)

Deprecated. See [Insight](#Insight).

### ❌ QueryRuleCustomData

[QueryRuleCustomData references](https://www.algolia.com/doc/api-reference/widgets/query-rule-custom-data/js/)

> You may want to use this widget to display banners or recommendations returned by Rules, and that match search parameters.

Reviewer: ```suggestion
You may want to use this widget to display banners or recommendations returned by rules, and that match search parameters.
```

Author: [Rules](https://www.algolia.com/doc/guides/managing-results/rules/detecting-intent/) are a feature of Algolia, should we not keep the capital letter in such cases?

Reviewer: Ok as you wish! It was not obvious when reading the README, but since there is a lot of things to read, I might have missed this explaination

No compatibility because MeiliSearch does not support [rules](https://www.algolia.com/doc/guides/managing-results/rules/rules-overview/).

### ❌ QueryRuleContext

[Query rule context references](https://www.algolia.com/doc/api-reference/widgets/query-rule-context/js/)

> The queryRuleContext widget lets you apply ruleContexts based on filters to trigger context-dependent Rules.

Reviewer: ```suggestion
The queryRuleContext widget lets you apply ruleContexts based on filters to trigger context-dependent rules.
```

Author: [Rules](https://www.algolia.com/doc/guides/managing-results/rules/detecting-intent/) are a feature of Algolia, should we not keep the capital letter in such cases?

No compatibility because MeiliSearch does not support [rules](https://www.algolia.com/doc/guides/managing-results/rules/rules-overview/).

### ❌ SortBy

[Sort by references](https://www.algolia.com/doc/api-reference/widgets/sort-by/js/)

The sortBy widget displays a list of indices, allowing a user to change the way hits are sorted (with replica indices). Another common use case is to let the user switch between different indices.

No compatibility because MeiliSearch does not support hierarchical facets.

> If you'd like to get the sort by features, please vote for it in the [roadmap]https://roadmap.meilisearch.com/c/32-sort-by?utm_medium=social&utm_source=portal_share).

Reviewer: ```suggestion
If you'd like to get the "SortBy" feature, please vote for it in the [roadmap]https://roadmap.meilisearch.com/c/32-sort-by?utm_medium=social&utm_source=portal_share).
```

### ❌ RelevantSort

[Relevant Sort references](https://www.algolia.com/doc/api-reference/widgets/relevant-sort/js/)

Virtual indices allow you to use Relevant sort, a sorting mechanism that favors relevancy over the attribute you’re sorting on.

### ✅ Routing

Routing is configured inside `instantSearch` component. Please refer [to the documentation](https://www.algolia.com/doc/api-reference/widgets/simple-state-mapping/js/) for further implementation information.

> ## 🤖 Compatibility with MeiliSearch and InstantSearch

Reviewer: Can you out this section before ``. I'm worried the users never reach this section since the README is huge.

Reviewer: Sorry, I wanted to say the section before "Example with InstantSearch"

- If you use Vue, check out [meilisearch-vue](https://github.com/meilisearch/meilisearch-vue/)
- If you use Angular, check out [meilisearch-angular](https://github.com/meilisearch/meilisearch-angular/)

## 🤖 Compatibility with MeiliSearch and InstantSearch

This package only guarantees: compatibility with:

> - The version v0.20.0 of [MeiliSearch](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.20.0).

Reviewer: Can you follow the same presentation as for strapi? It will make easier the update of this part of the README with meili-bot. The sentence should be the same 😇

- The version v4 of [InstantSearch.js](https://github.com/algolia/instantsearch.js/releases/tag/v4.23.0)

## API resources

> List of all the components that are available in [instantSearch](https://github.com/algolia/instantsearch.js) and their compatibilty with [MeiliSearch](https://github.com/meilisearch/meilisearch-react/).

Reviewer: ```suggestion
List of all the components that are available in [instantSearch](https://github.com/algolia/instantsearch.js) and their compatibilty with [MeiliSearch](https://github.com/meilisearch/MeiliSearch/).
```
